tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,358 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import base64
+import errno as errno
+import glob
+import random
+import string
+import tempfile
+import uuid
+
+from datetime import datetime, timedelta, tzinfo
+from azurelinuxagent.common.protocol import get_protocol_util
+import azurelinuxagent.common.utils.shellutil as shellutil
+from azurelinuxagent.common.protocol.wire import *
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.osutil.default import DefaultOSUtil
+from azurelinuxagent.common.utils.cryptutil import CryptUtil
+from azurelinuxagent.common.exception import CryptError
+from azurelinuxagent.ga.remoteaccess import RemoteAccessHandler, REMOTE_USR_EXPIRATION_FORMAT, MAX_TRY_ATTEMPT
+from tests.common.osutil.mock_osutil import MockOSUtil
+from tests.protocol.mockwiredata import WireProtocolData, DATA_FILE
+from tests.tools import *
+
+class TestRemoteAccessHandler(AgentTestCase):
+
+    # add_user tests
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
+        self.assertEqual(actual_user[4], "JIT Account")
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_bad_creation_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = ""
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(0, os_util.all_users.keys().__len__())
```

Comment:
At the very least, `self.assertEqual(0, len(os_util.all_users))`. But the architected interface for examining the list of users in the system is `os_util.get_users()`, which you're very careful to implement in mock_osutil. I would think your test code would thus be
```
self.assertEqual(0, len(os_util.get_users()))
```

Reply:
Updated to use the interface

---
In reply to: [194534879](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194534879) [](ancestors = 194534879)